zeppelin-web-angular/e2e/tests/notebook/published/published-paragraph.spec.ts

```diff
@@ -87,55 +96,135 @@ test.describe('Published Paragraph', () => {
+
+      const runModalButton = modal.locator('button:has-text("Run")');
+      await expect(runModalButton).toBeVisible();
+      await runModalButton.click();
+      await expect(modal).not.toBeVisible({ timeout: 10000 });
+
+      // Verify published container is present and ready (might be initially hidden)
+      await expect(publishedContainer).toBeAttached({ timeout: 10000 });
+
+      // Verify we're actually in published mode by checking page structure
+      const isPublishedMode = await page.evaluate(() => document.querySelector('zeppelin-publish-paragraph') !== null);
+      expect(isPublishedMode).toBe(true);
+    });
   });
 
-  test('should show confirmation modal and allow running the paragraph', async ({ page }) => {
-    const { noteId, paragraphId } = testNotebook;
+  test.describe('Published Mode Functionality', () => {
+    test('should hide editing controls in published mode', async ({ page }) => {
+      const { noteId, paragraphId } = testNotebook;
 
-    await publishedParagraphPage.navigateToNotebook(noteId);
+      await page.goto(`/#/notebook/${noteId}/paragraph/${paragraphId}`);
+      await page.waitForLoadState('networkidle');
 
-    const paragraphElement = page.locator('zeppelin-notebook-paragraph').first();
-    const paragraphResult = paragraphElement.locator('zeppelin-notebook-paragraph-result');
+      // In published mode, code editor and control panel should be hidden
+      const codeEditor = page.locator('zeppelin-notebook-paragraph-code-editor');
+      const controlPanel = page.locator('zeppelin-notebook-paragraph-control');
 
-    // Only clear output if result exists
-    if (await paragraphResult.isVisible()) {
-      const settingsButton = paragraphElement.locator('a[nz-dropdown]');
-      await settingsButton.click();
+      const isCodeEditorVisible = await codeEditor.isVisible();
+      const isControlPanelVisible = await controlPanel.isVisible();
 
-      const clearOutputButton = page.locator('li.list-item:has-text("Clear output")');
-      await clearOutputButton.click();
-      await expect(paragraphResult).toBeHidden();
-    }
+      if (isCodeEditorVisible) {
+        await expect(codeEditor).toBeHidden();
+      }
+      if (isControlPanelVisible) {
+        await expect(controlPanel).toBeHidden();
+      }
+    });
+
+    test('should display dynamic forms in published mode', async ({ page }) => {
+      const { noteId, paragraphId } = testNotebook;
 
-    await publishedParagraphPage.navigateToPublishedParagraph(noteId, paragraphId);
+      await page.goto(`/#/notebook/${noteId}/paragraph/${paragraphId}`);
+      await page.waitForLoadState('networkidle');
+
+      // Dynamic forms should be visible and functional in published mode
+      const isDynamicFormsVisible = await page.locator('zeppelin-notebook-paragraph-dynamic-forms').isVisible();
+      if (isDynamicFormsVisible) {
+        await expect(page.locator('zeppelin-notebook-paragraph-dynamic-forms')).toBeVisible();
+      }
```

Comment:
Just to clarify the intention here: if the dynamic forms are not visible, this check won’t fail the test.
Are we intentionally allowing that, or did we want to assert that dynamic forms must be visible in published mode?